referenceframe/urdf.go

```diff
@@ -0,0 +1,258 @@
+		// URDFs manage frame origins within joints instead of within links
+		geoXYZ := convStringAttrToFloats(jointElem.Origin.XYZ)
+		geoRPY := convStringAttrToFloats(jointElem.Origin.RPY)
+		jointPose := spatial.NewPoseFromOrientation(r3.Vector{X: geoXYZ[0], Y: geoXYZ[1], Z: geoXYZ[2]},
+			&spatial.EulerAngles{Roll: geoRPY[0], Pitch: geoRPY[1], Yaw: geoRPY[2]})
+		jointOrigins[jointName] = jointPose
+	}
+
+	// Handle links
+	terminalLinks := []string{}
+
+	for _, linkElem := range urdf.Links {
+		linkName := linkElem.Name
+
+		// TODO(wspies): Consider this link more carefully
+		if strings.Contains(linkName, "world") {
+			continue
+		}
+
+		// In a majority of cases, the end effector link will not have any joint listed as a child, nor will it have any
+		// associated spatial transforms.
+		// TODO(wspies): There should be only one of these terminal links, but some robots may have additional terminal
+		// links, so as to accomodate actuators, sensors, or other devices. Might not need to deal with that now...
+		if _, is_ok := childMap[linkName]; !is_ok {
+			terminalLinks = append(terminalLinks, linkName)
+			transforms[linkName], err = NewStaticFrame(linkName, zeroPose)
+			continue
+		}
+
+		// Otherwise, parse important details about each link, overriding where necessary
+		// The child joint origin element will have translation and rotation details for each link
+		linkPose := jointOrigins[childMap[linkName]]
+
+		if len(linkElem.Collision) > 0 {
+			// TODO(wspies): Finish up handling collision geometry
+			// Create the geometry for the collision object
+			//geoXYZ := convStringAttrToFloats(linkElem.Collision[0].Origin.XYZ)
+			//geoRPY := convStringAttrToFloats(linkElem.Collision[0].Origin.RPY)
+			//linkPosition = r3.Vector{X: geoXYZ[0], Y: geoXYZ[1], Z: geoXYZ[2]}
+			//linkOrientation = &spatial.EulerAngles{Roll: geoRPY[0], Pitch: geoRPY[1], Yaw: geoRPY[2]}
+			//linkPose = spatial.NewPoseFromOrientation(linkPosition, linkOrientation)
+
+			// TODO(wspies): Decide what to do about box versus sphere
+			//var geometryCreator spatial.GeometryCreator
+			//if box {
+			//	geometryCreator = spatial.NewBoxCreator()
+			//	transforms[linkName], err = NewStaticFrameWithGeometry(linkName, linkPose, geometryCreator)
+			//} else if sphere {
+			//	geometryCreator = spatial.NewSphereCreator()
+			//	transforms[linkName], err = NewStaticFrameWithGeometry(linkName, linkPose, geometryCreator)
+			//} else {
+			//	err = errors.Errorf("Unsupported collision geometry type detected for [ %v ] link", linkElem.Collision[0].Name)
+			//}
+			fmt.Println("Working with collision object...")  // DEBUG(wspies): Remove after testing
+		} else {
+			transforms[linkName], err = NewStaticFrame(linkName, linkPose)
+		}
+	}
+
+	if len(terminalLinks) != 1 {
+		return nil, errors.Errorf("Invalid terminal link count: %d", len(terminalLinks))
+	}
+
+	// Create joint and link ordering, starting with the end effector link and going backwards
```

Comment:
looks like this is mostlty the same as the code in `model_json` might be possible to make a helper function